engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,281 @@
+	s.dataProviderFactory = dpmock.NewDataProviderFactory(s.T())
+}
+
+// TestConfigureConnection ensures that the WebSocket connection is configured correctly.
+func (s *ControllerSuite) TestConfigureConnection() {
+	controller := s.initializeController()
+
+	// Mock configureConnection to succeed
+	s.mockConnectionSetup()
+
+	// Call configureConnection and check for errors
+	err := controller.configureConnection()
+	s.Require().NoError(err, "configureConnection should not return an error")
+
+	// Assert expectations
+	s.connection.AssertExpectations(s.T())
+}
+
+// TestControllerShutdown ensures that HandleConnection shuts down gracefully when an error occurs.
+func (s *ControllerSuite) TestControllerShutdown() {
+	s.T().Run("keepalive routine failed", func(*testing.T) {
+		controller := s.initializeController()
+
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+
+		// Mock keepalive to return an error
+		done := make(chan struct{}, 1)
+		s.connection.On("WriteControl", websocket.PingMessage, mock.Anything).Return(func(int, time.Time) error {
+			close(done)
+			return websocket.ErrCloseSent
+		}).Once()
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Return(func(interface{}) error {
+				_, ok := <-done
+				if !ok {
+					return websocket.ErrCloseSent
+				}
+				return nil
+			}).
+			Once()
+
+		s.connection.On("Close").Return(nil).Once()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		controller.HandleConnection(ctx)
+
+		s.Require().True(controller.shutdown.Load())
+		// Ensure all expectations are met
+		s.connection.AssertExpectations(s.T())
+	})
+
+	s.T().Run("read routine failed", func(*testing.T) {
+		controller := s.initializeController()
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Return(func(_ interface{}) error {
+				return assert.AnError
+			}).
+			Once()
+
+		s.connection.On("Close").Return(nil).Once()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		controller.HandleConnection(ctx)
+
+		s.Require().True(controller.shutdown.Load())
+		// Ensure all expectations are met
+		s.connection.AssertExpectations(s.T())
+	})
+
+	s.T().Run("write routine failed", func(*testing.T) {
+		controller := s.initializeController()
+
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+		blocksDataProvider := s.mockBlockDataProviderSetup(uuid.New())
+
+		done := make(chan struct{}, 1)
+		requestMessage := models.SubscribeMessageRequest{
+			BaseMessageRequest: models.BaseMessageRequest{Action: "subscribe"},
+			Topic:              "blocks",
+			Arguments:          nil,
+		}
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Run(func(args mock.Arguments) {
+				reqMsg, ok := args.Get(0).(*json.RawMessage)
+				s.Require().True(ok)
+				msg, err := json.Marshal(requestMessage)
+				s.Require().NoError(err)
+				*reqMsg = msg
```

Comment:
what does this do? isn't `reqMsg` a local variable that's thrown away at the end of the function?

Reply:
The idea of this code is to update the first argument of `ReadJSON`. Even though `reqMsg` is a local variable, it serves as a pointer to the argument passed into the mock. By modifying *`reqMsg`, the code changes the actual memory location that the pointer refers to.

Reply:
why is this needed vs passing the info in somehow? Can you add some comments explaining. it's unclear to me so it will most likely be unclear to the next person that works on these tests.

Reply:
let's do it like that
```
	conn.On("ReadJSON", requestMessage).
		Return(nil).
		Once()
```

Reply:
> let's do it like that
> 
> ```
> 	conn.On("ReadJSON", requestMessage).
> 		Return(nil).
> 		Once()
> ```
This will not be worked cause, the strict pointer comparison by the mock library expects the exact pointer passed during the mock setup to be the same as the pointer passed during runtime. Since these pointers are not the same, the mock is not triggered.

As I understood, to properly mock the `ReadJSON` call, there are 2 options:
1) Use `mock.Anything` as argument and then modify the argument inside the `Run()` callback. (This approach is implemented in current test)
2) Use mock.`MatchedBy`, something like 
		
```
s.connection.On("ReadJSON", mock.MatchedBy(func(arg interface{}) bool {
			rawMsg, ok := arg.(*json.RawMessage)
			if !ok {
				return false
			}
			*rawMsg = msg
			return true
		})).
			Return(nil).
			Once()
```
I’ve added a detailed comment in the code explaining why `ReadJSON` is mocked this way